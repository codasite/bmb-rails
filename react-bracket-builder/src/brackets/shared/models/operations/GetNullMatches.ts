export const getNullMatches = (numRounds: number): null[][] => {
  let rounds: any[] = []
  for (let i = numRounds - 1; i >= 0; i--) {
    rounds.push(new Array(Math.pow(2, i)).fill(null))
  }
  return rounds
}
