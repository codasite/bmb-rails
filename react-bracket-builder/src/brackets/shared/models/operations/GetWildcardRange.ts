import { WildcardPlacement } from '../WildcardPlacement'
import { WildcardRange } from '../WildcardRange'

export const getWildcardRange = (
  start: number,
  end: number,
  count: number,
  placement: WildcardPlacement
): WildcardRange[] => {
  switch (placement) {
    case WildcardPlacement.Top:
      return [{ min: start, max: start + count }]
    case WildcardPlacement.Bottom:
      return [{ min: end - count, max: end }]
    case WildcardPlacement.Center:
      const offset = (end - start - count) / 2
      return [{ min: start + offset, max: end - offset }]
    case WildcardPlacement.Split:
      return [
        { min: start, max: start + count / 2 },
        { min: end - count / 2, max: end },
      ]
  }
}
