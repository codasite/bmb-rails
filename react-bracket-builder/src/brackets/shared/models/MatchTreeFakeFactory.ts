import { MatchTree } from './MatchTree'

export function fourTeamMatchTree() {
  const matches = [
    [
      {
        id: 9,
        roundIndex: 0,
        matchIndex: 0,
        team1: { id: 17, name: 'Team 1' },
        team2: { id: 18, name: 'Team 2' },
        team1Wins: true,
      },
      {
        id: 10,
        roundIndex: 0,
        matchIndex: 1,
        team1: { id: 19, name: 'Team 3' },
        team2: { id: 20, name: 'Team 4' },
        team1Wins: true,
      },
    ],
    [{ roundIndex: 1, matchIndex: 1, team2Wins: true }],
  ]
  return MatchTree.deserialize({ rounds: matches })
}
