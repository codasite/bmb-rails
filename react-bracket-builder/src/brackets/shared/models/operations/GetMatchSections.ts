import { Round } from '../Round'
import { Nullable } from '../../../../utils/types'
import { MatchNode } from './MatchNode'

export function getFirstMatches(rounds: Round[]): MatchNode[] {
  if (rounds.length < 7) {
    return []
  }
  return rounds[0].matches.filter((match) => match !== null)
}

export function getSideMatches(rounds: Round[]): {
  left: Nullable<MatchNode>[][]
  right: Nullable<MatchNode>[][]
} {
  const sideRounds = rounds.slice(rounds.length < 7 ? 0 : 1, rounds.length - 1)
  return sideRounds.reduce(
    (matches, round) => ({
      left: [...matches.left, round.matches.slice(0, round.matches.length / 2)],
      right: [round.matches.slice(round.matches.length / 2), ...matches.right],
    }),
    { left: [], right: [] }
  )
}

export function getFinalMatches(rounds: Round[]): Nullable<MatchNode>[] {
  return rounds[rounds.length - 1].matches
}
