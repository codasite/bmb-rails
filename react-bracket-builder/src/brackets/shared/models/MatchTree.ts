import { Nullable } from '../../../utils/types'
import { BracketRes, MatchPick, MatchReq, MatchTreeRepr } from '../api'
import { WildcardPlacement } from './WildcardPlacement'
import { Team } from './Team'
import { MatchNode } from './operations/MatchNode'
import { Round } from './Round'
import { matchReprFromNumTeams } from './operations/MatchReprFromNumTeams'
import { getNumRounds } from './operations/GetNumRounds'
import { matchReprFromRes } from './operations/MatchReprFromRes'
import { linkNodes } from './operations/LinkNodes'

export class MatchTree {
  rounds: Round[]
  private numTeams: number
  private wildcardPlacement?: WildcardPlacement
  isVoting: boolean
  liveRoundIndex: number

  constructor(
    rounds: Round[] = [],
    wildcardPlacement?: WildcardPlacement,
    isVoting: boolean = false,
    liveRoundIndex: number = 0
  ) {
    linkNodes(rounds)
    this.rounds = rounds
    this.wildcardPlacement = wildcardPlacement
    this.isVoting = isVoting
    this.liveRoundIndex = liveRoundIndex
  }

  getRootMatch(): Nullable<MatchNode> {
    const lastRound = this.rounds[this.rounds.length - 1]
    if (!lastRound) {
      return null
    }
    return lastRound.matches[0]
  }

  clone(): MatchTree {
    return MatchTree.deserialize(this.serialize())
  }

  equals(other: MatchTree): boolean {
    return (
      JSON.stringify(this.serialize()) === JSON.stringify(other.serialize())
    )
  }

  /**
   * Returns the total number of POSSIBLE teams in the tournament.
   * This should reflect the number of teams that would be used to create the same structure with MatchTree.fromNumTeams()
   */
  getNumTeams(): number {
    if (this.numTeams) {
      return this.numTeams
    }
    // count all non null matches in the tree
    const numMatches = this.rounds.reduce((acc, round) => {
      return (
        acc +
        round.matches.reduce((acc, match) => {
          if (match) {
            return acc + 1
          }
          return acc
        }, 0)
      )
    }, 0)

    const numTeams = numMatches + 1
    this.numTeams = numTeams

    return numTeams
  }

  getWildcardPlacement(): WildcardPlacement | undefined {
    return this.wildcardPlacement
  }

  serialize(): MatchTreeRepr {
    const tree = this
    const rounds = tree.rounds.map((round) => {
      return round.serialize()
    })
    return {
      rounds: rounds,
      wildcardPlacement: tree.wildcardPlacement,
      isVoting: tree.isVoting,
      liveRoundIndex: tree.liveRoundIndex,
    }
  }

  advanceTeam = (
    roundIndex: number,
    matchIndex: number,
    left: boolean,
    requireComplete: boolean = false
  ) => {
    if (requireComplete && roundIndex > 0) {
      const prevRound = this.rounds[roundIndex - 1]
      if (prevRound && !prevRound.allPicked()) {
        return
      }
    }
    const round = this.rounds[roundIndex]
    const match = round.matches[matchIndex]
    if (!match) {
      return
    }
    match.team1Wins = false
    match.team2Wins = false
    if (left) {
      match.team1Wins = true
    } else {
      match.team2Wins = true
    }
  }

  anyPicked = (): boolean => {
    return this.rounds.some((round) => {
      return round.matches.some((match) => {
        if (!match) {
          return false
        }
        return match.team1Wins || match.team2Wins
      })
    })
  }

  allPicked = (): boolean => {
    if (this.isVoting) {
      return this.rounds[this.liveRoundIndex].allPicked()
    }
    return this.rounds.every((round) => {
      return round.allPicked()
    })
  }

  allTeamsAdded = (): boolean => {
    return this.everyMatch((match) => {
      let hasNeededTeams = true
      if (!match.left && !match.getTeam1()?.name) {
        hasNeededTeams = false
      }
      if (hasNeededTeams && !match.right && !match.getTeam2()?.name) {
        hasNeededTeams = false
      }
      return hasNeededTeams
    })
  }

  toMatchReq = (): MatchReq[] => {
    const matches: MatchReq[] = []
    this.forEachMatch((match, matchIndex, roundIndex) => {
      const matchReq: MatchReq = {
        roundIndex: match.roundIndex,
        matchIndex: match.matchIndex,
      }
      if (!match.left) {
        matchReq.team1 = match.getTeam1()?.toTeamReq()
      }
      if (!match.right) {
        matchReq.team2 = match.getTeam2()?.toTeamReq()
      }
      if (matchReq.team1 || matchReq.team2) {
        matches.push(matchReq)
      }
    })
    return matches
  }

  toMatchPicks = (): MatchPick[] => {
    const picks: MatchPick[] = []
    this.rounds.forEach((round, roundIndex) => {
      round.matches.forEach((match, matchIndex) => {
        if (!match) {
          return
        }
        const { team1Wins, team2Wins } = match
        const team1 = match.getTeam1()
        const team2 = match.getTeam2()
        const team1Winner = team1Wins && team1
        const team2Winner = team2Wins && team2

        if (team1Winner && team1.id) {
          picks.push({
            roundIndex,
            matchIndex,
            winningTeamId: team1.id,
          })
        } else if (team2Winner && team2.id) {
          picks.push({
            roundIndex,
            matchIndex,
            winningTeamId: team2.id,
          })
        }
      })
    })
    return picks
  }

  forEachMatch = (
    callback: (match: MatchNode, matchIndex: number, roundIndex: number) => void
  ) => {
    this.rounds.forEach((round, roundIndex) => {
      round.matches.forEach((match, matchIndex) => {
        if (!match) {
          return
        }
        callback(match, matchIndex, roundIndex)
      })
    })
  }

  everyMatch = (
    callback: (
      match: MatchNode,
      matchIndex: number,
      roundIndex: number
    ) => boolean
  ) => {
    return this.rounds.every((round, roundIndex) => {
      return round.matches.every((match, matchIndex) => {
        if (!match) {
          return true
        }
        return callback(match, matchIndex, roundIndex)
      })
    })
  }

  findMatch = (
    callback: (match: Nullable<MatchNode>) => boolean
  ): Nullable<MatchNode> => {
    let foundMatch: Nullable<MatchNode> = null
    this.rounds.some((round) => {
      return round.matches.some((match) => {
        if (callback(match)) {
          foundMatch = match
          return true
        }
        return false
      })
    })
    return foundMatch
  }

  static fromNumTeams(
    numTeams: number,
    wildcardPlacement: WildcardPlacement = WildcardPlacement.Top
  ): MatchTree {
    const matches = matchReprFromNumTeams(numTeams, wildcardPlacement)
    return MatchTree.deserialize({
      rounds: matches,
      wildcardPlacement,
    })
  }

  static fromMatchRes(
    bracket: Pick<
      BracketRes,
      'matches' | 'numTeams' | 'isVoting' | 'liveRoundIndex'
    >,
    wildcardPlacement?: WildcardPlacement
  ): MatchTree | null {
    const numRounds = getNumRounds(bracket.numTeams)

    const nestedMatches = matchReprFromRes(numRounds, bracket.matches)
    return MatchTree.deserialize({
      rounds: nestedMatches,
      wildcardPlacement,
      isVoting: bracket.isVoting,
      liveRoundIndex: bracket.liveRoundIndex,
    })
  }

  static fromPicks(
    bracket: Pick<
      BracketRes,
      'matches' | 'numTeams' | 'isVoting' | 'liveRoundIndex'
    >,
    picks: MatchPick[],
    wildcardPlacement?: WildcardPlacement
  ): MatchTree | null {
    const matchTree = MatchTree.fromMatchRes(bracket, wildcardPlacement)
    if (!matchTree) {
      return null
    }
    for (const pick of picks) {
      const { roundIndex, matchIndex, winningTeamId } = pick
      const match = matchTree.rounds[roundIndex].matches[matchIndex]
      if (!match) {
        throw new Error('Match not found')
      }
      match._pick = pick

      const team1 = match.getTeam1()
      const team2 = match.getTeam2()
      if (team1?.id === winningTeamId) {
        match.team1Wins = true
      } else if (team2?.id === winningTeamId) {
        match.team2Wins = true
      } else {
        throw new Error('Match does not have winningTeam from pick')
      }
    }
    return matchTree
  }

  static deserialize(matchTreeRepr: MatchTreeRepr): MatchTree {
    const { rounds: matchRes, wildcardPlacement } = matchTreeRepr
    const rounds = matchRes.map((round, roundIndex) => {
      const depth = matchRes.length - roundIndex - 1
      const newRound = new Round(roundIndex, depth)
      const matches = round.map((match, matchIndex) => {
        if (match === null) {
          return null
        }
        const {
          id,
          team1: team1Repr,
          team2: team2Repr,
          team1Wins,
          team2Wins,
          pick,
        } = match
        const team1 = team1Repr ? new Team(team1Repr.name, team1Repr.id) : null
        const team2 = team2Repr ? new Team(team2Repr.name, team2Repr.id) : null
        // const newMatch = new MatchNode(roundIndex, matchIndex, depth, match.id, team1, team2, team1Wins, team2Wins)
        const newMatch = new MatchNode({
          id,
          matchIndex,
          roundIndex,
          team1,
          team2,
          team1Wins,
          team2Wins,
          left: null,
          right: null,
          parent: null,
          depth,
          pick,
        })
        return newMatch
      })
      newRound.matches = matches
      return newRound
    })
    const tree = new MatchTree(
      rounds,
      wildcardPlacement,
      matchTreeRepr.isVoting,
      matchTreeRepr.liveRoundIndex
    )
    return tree
  }
}
