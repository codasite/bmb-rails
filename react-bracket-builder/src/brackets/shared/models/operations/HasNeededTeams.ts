import { MatchNode } from './MatchNode'

export const hasNeededTeams = (match: MatchNode): boolean => {
  let hasNeededTeams = true
  if (!match) {
    return true
  }
  if (!match.left && !match.getTeam1()?.name) {
    hasNeededTeams = false
  }
  if (hasNeededTeams && !match.right && !match.getTeam2()?.name) {
    hasNeededTeams = false
  }
  return hasNeededTeams
}
