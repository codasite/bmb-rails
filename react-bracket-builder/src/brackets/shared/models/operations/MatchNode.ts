import { Nullable } from '../../../../utils/types'
import { Team } from '../Team'
import { MatchNodeArgs } from './MatchNodeArgs'
import { MatchPick, MatchRepr } from '../../api/types/bracket'
import { TeamPosition } from '../../components/types'

export class MatchNode {
  id?: number
  matchIndex: number
  roundIndex: number
  private team1: Nullable<Team> = null
  private team2: Nullable<Team> = null
  team1Wins?: boolean = false
  team2Wins?: boolean = false
  left: Nullable<MatchNode> = null
  right: Nullable<MatchNode> = null
  parent: Nullable<MatchNode> = null
  depth: number
  _pick: MatchPick | null = null
  constructor(args: MatchNodeArgs) {
    const {
      id,
      matchIndex,
      roundIndex,
      team1,
      team2,
      team1Wins,
      team2Wins,
      left,
      right,
      parent,
      depth,
      pick,
    } = args
    this.matchIndex = matchIndex
    this.roundIndex = roundIndex
    this.depth = depth
    this.id = id
    this.team1Wins = team1Wins
    this.team2Wins = team2Wins
    this.left = left ? left : null
    this.right = right ? right : null
    this.parent = parent ? parent : null
    this.team1 = team1 ? team1 : null
    this.team2 = team2 ? team2 : null
    this._pick = pick ? pick : null
  }
  serialize(): MatchRepr {
    const { id, matchIndex, roundIndex, team1, team2, team1Wins, team2Wins } =
      this
    return {
      id,
      matchIndex,
      roundIndex,
      team1: team1 ? team1.serialize() : undefined,
      team2: team2 ? team2.serialize() : undefined,
      team1Wins,
      team2Wins,
      pick: this._pick,
    }
  }
  getWinner(): Nullable<Team> {
    if (this.team1Wins) {
      return this.getTeam1()
    } else if (this.team2Wins) {
      return this.getTeam2()
    }
    return null
  }
  isLeftChild(): boolean {
    return this.parent !== null && this.parent.left === this
  }
  isRightChild(): boolean {
    return !this.isLeftChild()
  }
  getTeam(position: TeamPosition): Nullable<Team> {
    if (position === 'left') {
      return this.getTeam1()
    } else if (position === 'right') {
      return this.getTeam2()
    }
    return null
  }
  getTeam1(): Nullable<Team> {
    return this.team1 ?? this.left?.getWinner()
  }
  getTeam2(): Nullable<Team> {
    return this.team2 ?? this.right?.getWinner()
  }
  setTeam1(team: Team): void {
    this.team1 = team
  }
  setTeam2(team: Team): void {
    this.team2 = team
  }
  setTeam(team: Team, isTeam1: boolean): void {
    if (isTeam1) {
      this.setTeam1(team)
    } else {
      this.setTeam2(team)
    }
  }
  pick(team: Nullable<Team>): void {
    if (!team) {
      return
    }
    this.team1Wins = false
    this.team2Wins = false
    if (this.getTeam1() === team) {
      this.team1Wins = true
    } else if (this.getTeam2() === team) {
      this.team2Wins = true
    }
  }
  isPicked(): boolean {
    return this.team1Wins || this.team2Wins
  }
  isLeafTeam(teamPosition: TeamPosition): boolean {
    if (teamPosition === 'left') {
      return this.left === null
    } else if (teamPosition === 'right') {
      return this.right === null
    }
    return false
  }
  hasLeafTeam(): boolean {
    return this.isLeafTeam('left') || this.isLeafTeam('right')
  }
}
