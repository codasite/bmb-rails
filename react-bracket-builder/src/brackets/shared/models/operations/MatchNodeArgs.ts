import { Nullable } from '../../../../utils/types'
import { MatchPick } from '../../api'
import { Team } from '../Team'
import { MatchNode } from './MatchNode'

export interface MatchNodeArgs {
  id?: number
  matchIndex: number
  roundIndex: number
  team1?: Nullable<Team>
  team2?: Nullable<Team>
  team1Wins?: boolean
  team2Wins?: boolean
  left?: Nullable<MatchNode>
  right?: Nullable<MatchNode>
  parent?: Nullable<MatchNode>
  depth: number
  pick?: MatchPick
}
