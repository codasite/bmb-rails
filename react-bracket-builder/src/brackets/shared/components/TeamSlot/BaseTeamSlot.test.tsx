// DO NOT REMOVE REACT IMPORT. Needed for image generator
import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { BaseTeamSlot } from './BaseTeamSlot' // Import your component
import '@testing-library/jest-dom/jest-globals' // for additional matchers like toBeInTheDocument
import { MatchTree } from '../../models/MatchTree'
import { MatchNode } from '../../models/operations/MatchNode'
import { Team } from '../../models/Team'
import { userEvent } from '@testing-library/user-event'

// Mock some functions and props
const mockOnTeamClick = jest.fn()
const team = new Team('Team A')
const match = new MatchNode({ matchIndex: 2, roundIndex: 1, depth: 1 })
const matchTree = MatchTree.fromNumTeams(8)

describe('BaseTeamSlot', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  test('renders with correct text and styles', () => {
    render(
      <BaseTeamSlot
        team={team}
        match={match}
        teamPosition="left"
        height={50}
        width={115}
        onTeamClick={mockOnTeamClick}
        matchTree={matchTree}
        placeholder="Placeholder"
      />
    )

    // Check if the text is rendered
    expect(screen.getByText('Team A')).toBeInTheDocument()

    // Check if the element has the correct data-testid
    const teamSlot = screen.getByTestId('team-slot-round-1-match-2-left')
    expect(teamSlot).toBeInTheDocument()

    // Verify the base styles are applied
    expect(teamSlot).toMatchSnapshot()
  })

  test('calls onTeamClick when clicked', () => {
    render(
      <BaseTeamSlot
        team={team}
        match={match}
        teamPosition="left"
        height={50}
        width={115}
        onTeamClick={mockOnTeamClick}
        matchTree={matchTree}
      />
    )

    const teamSlot = screen.getByTestId('team-slot-round-1-match-2-left')

    // Simulate a click event
    fireEvent.click(teamSlot)

    // Ensure the click handler was called
    expect(mockOnTeamClick).toHaveBeenCalledTimes(1)
    expect(mockOnTeamClick).toHaveBeenCalledWith(match, 'left', team)
  })

  test('does not call onTeamClick when teamClickDisabled', () => {
    const teamClickDisabled = jest.fn(() => true)

    render(
      <BaseTeamSlot
        team={team}
        match={match}
        teamPosition="left"
        height={50}
        width={115}
        onTeamClick={mockOnTeamClick}
        matchTree={matchTree}
        teamClickDisabled={teamClickDisabled}
      />
    )

    const teamSlot = screen.getByTestId('team-slot-round-1-match-2-left')

    // Simulate a click event
    fireEvent.click(teamSlot)

    // Ensure the click handler was not called
    expect(mockOnTeamClick).not.toHaveBeenCalled()
  })

  test('renders placeholder if no team name is provided', () => {
    render(
      <BaseTeamSlot
        match={match}
        teamPosition="left"
        height={50}
        width={115}
        placeholder="No Team"
        matchTree={matchTree}
      />
    )

    // Check if the placeholder is rendered
    expect(screen.getByText('No Team')).toBeInTheDocument()
  })
})
