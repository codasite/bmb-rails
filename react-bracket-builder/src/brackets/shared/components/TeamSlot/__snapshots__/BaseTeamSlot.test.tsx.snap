// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTeamSlot renders with correct text and styles 1`] = `
<button
  class="wpbb-team-1-2-left tw-relative tw-flex tw-justify-center tw-items-center tw-whitespace-nowrap tw-leading-none tw-uppercase tw-font-sans tw-w-[115px] tw-h-[50px] tw-text-white tw-font-500 tw-bg-transparent tw-cursor-pointer"
  data-testid="team-slot-round-1-match-2-left"
  style="border-width: 0px;"
>
  <span
    style="font-size: 16px; transform: scale(1); visibility: hidden;"
  >
    Team A
  </span>
</button>
`;
